Tests/MongoSwiftSyncTests/UnifiedTestRunner/Matching.swift

```diff
@@ -0,0 +1,311 @@
+        for (actualElt, expectedElt) in zip(actualElts, expectedArray) {
+            guard try matchesInner(expected: expectedElt, actual: actualElt, entities: entities) else {
+                return false
+            }
+        }
+
+        return true
+    case .int32, .int64, .double:
+        return matchesNumber(expected: expected, actual: actual)
+    default:
+        // if we made it here, the expected value is a non-document, non-array BSON, so we should expect `actual` to be
+        // a scalar value too.
+        guard case let .scalar(bson) = actual else {
+            return false
+        }
+        return bson == expected
+    }
+}
+
+/// When comparing numeric types (excluding Decimal128), test runners MUST consider 32-bit, 64-bit, and floating point
+/// numbers to be equal if their values are numerically equivalent.
+func matchesNumber(expected: BSON, actual: MatchableResult) -> Bool {
+    guard case let .scalar(bson) = actual else {
+        return false
+    }
+    guard let actualDouble = bson.toDouble() else {
+        return false
+    }
+
+    // fuzzy equals in case of e.g. rounding errors
+    return abs(actualDouble - expected.toDouble()!) < 0.0001
+}
+
+extension BSONDocument {
+    /// Returns whether this document is a special matching operator.
+    var isSpecialOperator: Bool {
+        self.count == 1 && self.keys[0].starts(with: "$$")
+    }
+}
+
+/// Determines whether `actual` satisfies the special matching operator in the provided `operatorDoc`.
+func matchesSpecial(operatorDoc: BSONDocument, actual: MatchableResult, entities: EntityMap) throws -> Bool {
+    let (op, value) = operatorDoc.first!
+    switch op {
+    case "$$exists":
+        let shouldExist = value.boolValue!
+        switch actual {
+        case .none:
+            return !shouldExist
+        default:
+            return shouldExist
+        }
+    case "$$type":
+        return matchesType(expectedType: value, actual: actual)
+    case "$$matchesEntity":
+        let id = value.stringValue!
+        let entity = try entities.getEntity(id: id).asBSON()
+        return try matchesInner(expected: entity, actual: actual, entities: entities)
+    case "$$matchesHexBytes":
+        throw TestError(message: "Unsupported special operator $$matchesHexBytes")
+    case "$$unsetOrMatches":
+        if case .none = actual {
+            return true
+        }
+        return try matchesInner(expected: value, actual: actual, entities: entities)
+    case "$$sessionLsid":
+        guard case let .document(actualDoc) = actual else {
+            return false
+        }
+        let id = value.stringValue!
+        let session = try entities.getEntity(id: id).asSession()
+        return actualDoc == session.id
+    default:
+        fatalError("Unrecognized special operator \(op)")
+    }
+}
+
+/// Determines whether `actual` satisfies the $$type operator value `expectedType`.
+func matchesType(expectedType: BSON, actual: MatchableResult) -> Bool {
```

Comment:
I realized that the value for $$type can be either a single string or an array of strings so I've updated the type matching code to reflect that.